ENH: allow int sequences as shape arguments in numpy.memmap

JanukanS opened the pull request:
The constructor for memmap objects initially required the shape argument to be a tuple, where non-tuple shape arguments would be made into a single member tuple. This behaviour was intended to allow an integer shape arguments. However this caused list or numpy.array shape arguments to raise errors within the constructor. Modifications made to the shape argument handling within the constructor now allow for integer sequences to be used for shape. A test was added and constructor docstring was updated to account for these changes.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

JanukanS committed:
ENH: allow int sequences as shape arguments in numpy.memmap

The constructor for memmap objects initially required the shape argument
to be a tuple, where non-tuple shape arguments would be made into a
single member tuple. This behaviour was intended to allow an integer
shape arguments. However this caused list or numpy.array shape arguments
to raise errors within the constructor. Modifications made to the
shape argument handling within the constructor now allow for
integer sequences to be used for shape. A test was added and
constructor docstring was updated to account for these changes.

JanukanS committed:
ENH: allow int sequences as shape arguments in numpy.memmap

The constructor for memmap objects initially required the shape argument
to be a tuple, where non-tuple shape arguments would be made into a
single member tuple. This behaviour was intended to allow an integer
shape arguments. However this caused list or numpy.array shape arguments
to raise errors within the constructor. Modifications made to the
shape argument handling within the constructor now allow for
integer sequences to be used for shape. A test was added and
constructor docstring was updated to account for these changes.

JanukanS committed:
Shape Argument Always Cast to Tuple

seberg commented on numpy/core/memmap.py:
```diff
@@ -242,8 +242,10 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
                 shape = (size,)
             else:
-                if not isinstance(shape, tuple):
+                if not hasattr(shape, '__iter__'):
```
I would suggest using the same pattern as `normalize_axis_tuple` (I think we may have another pattern around).  You are adding `int` support, so should have a `.. versionchanged::` tag (and release note if you like).

(That will allow `np.array(5)` as a shape.  Not that that should matter in practice, just think it may make sense to reuse a pattern we use in a few places.)

The tuple check is fine, but I suspect not worth it if it is already a tuple calling `tuple` is a no-op (except for subclasses maybe).

JanukanS committed:
Consistent type handling

Type handling within np.memmap constructor now matches the type
handling pattern found in normalize_axis_tuple. The test for
type handling was tested with 4 cases. Within the documentation,
the shape argument now has a version changed tag. A release note
has been added for this improvement.

JanukanS commented:
Just a clarification, the versionchanged tag should be in reference to version 1.25?

seberg commented on numpy/core/memmap.py:
```diff
@@ -75,12 +75,17 @@ class memmap(ndarray):
         will be 1-D with the number of elements determined by file size
         and data-type.
+
+        .. versionchanged:: 1.25
```
```suggestion
        .. versionchanged:: 2.0
```
Sorry, was ready early enough really, but it slipped now.

seberg commented on numpy/core/tests/test_memmap.py:
```diff
@@ -6,7 +6,9 @@
 
 from numpy import (
-    memmap, sum, average, prod, ndarray, isscalar, add, subtract, multiply)
+    memmap, sum, average, prod, ndarray, isscalar, add, subtract, multiply, 
+    array
+    )
```
```suggestion
    memmap, sum, average, prod, ndarray, isscalar, add, subtract, multiply)
```

seberg commented on numpy/core/tests/test_memmap.py:
```diff
@@ -213,3 +215,9 @@ def test_empty_array(self):
+        memmap(self.tmpfp, shape=3, mode='w+')
+        memmap(self.tmpfp, shape=self.shape, mode='w+')
+        memmap(self.tmpfp, shape=list(self.shape), mode='w+')
+        memmap(self.tmpfp, shape=array(self.shape), mode='w+')
```
```suggestion
        memmap(self.tmpfp, shape=asarray(self.shape), mode='w+')
```
(I would generally prefer to just use `np.` but the file uses it already.)

seberg commented on numpy/core/memmap.py:
```diff
@@ -22,7 +22,7 @@
 class memmap(ndarray):
     """Create a memory-map to an array stored in a *binary* file on disk.
-
+    
```
```suggestion

```

seberg reviewed (approved):
Thanks, LGTM, sorry that it slipped the release.  Will commit those few small nitpicky fixups.

seberg committed:
Apply suggestions from code review

seberg commented on numpy/core/memmap.py:
```diff
@@ -22,7 +22,6 @@
 class memmap(ndarray):
     """Create a memory-map to an array stored in a *binary* file on disk.
-
     Memory-mapped files are used for accessing small segments of large files
```
```suggestion

    Memory-mapped files are used for accessing small segments of large files
```
Hah, github didn't like the empty line...

seberg committed:
Apply suggestions from code review

seberg commented:
Thanks @JanukanS.
